Src/Plugins/AzureDevOpsConfiguration/AZC102_190.DoNotGrantAllPipelinesAccessValidator.cs

```diff
@@ -0,0 +1,154 @@
+            {
+                return ValidationResult.CreateNoMatch();
+            }
+
+            var validationResult = new ValidationResult
+            {
+                Fingerprint = new Fingerprint
+                {
+                    Host = org.Value.String,
+                    Resource = project.Value.String,
+                    Id = serviceConnectionId.Value.String,
+                    Platform = nameof(AssetPlatform.AzureDevOps),
+                },
+                ValidationState = ValidationState.Unknown,
+            };
+            return new[] { validationResult };
+        }
+
+        protected override ValidationState IsValidDynamicHelper(ref Fingerprint fingerprint,
+                                                                ref string message,
+                                                                IDictionary<string, string> options,
+                                                                ref ResultLevelKind resultLevelKind)
+        {
+            try
+            {
+                string organization = fingerprint.Host;
+                string project = fingerprint.Resource;
+                string serviceConnectionId = fingerprint.Id;
+
+                adoPat ??= ReadPatFromFile(AdoPatFile);
+                if (string.IsNullOrEmpty(adoPat))
+                {
+                    return ValidationState.Unauthorized;
+                }
+
+                HttpClient httpClient = CreateOrRetrieveCachedHttpClient();
+                httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+                httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic",
+                    Convert.ToBase64String(
+                        ASCIIEncoding.ASCII.GetBytes(
+                            string.Format("{0}:{1}", string.Empty, adoPat))));
+
+                // since DynamicValidatorBase already has a cache for <fingerprint, result>
+                // IsValidDynamicHelper will not be called if same fingerprint combinations,
+                // do not need cache for same organization/project/seviceconnectionid combination
+                using HttpResponseMessage response = httpClient.GetAsync(
+                    string.Format(PipelinePermissionAPI, organization, project, serviceConnectionId))
+                    .GetAwaiter()
+                    .GetResult();
+
+                switch (response.StatusCode)
+                {
+                    case HttpStatusCode.OK:
+                    {
+                        return VerifyResponse(
+                            response.Content.ReadAsStringAsync().GetAwaiter().GetResult(),
+                            ref message);
+                    }
+
+                    case HttpStatusCode.Unauthorized:
+                    case HttpStatusCode.Forbidden:
+                    {
+                        message = NotAuthorizedMessage;
```

Comment:
The reason to override this message is when status code is unauthorized/forbidden, a result level is information will generated. with this message, the result message will be more specific:

e.g. 
with the message.
[Service connection Id:AZC102.serviceconnection.DotNotShareWithAllPipelines.json]() in organization 'testorg.visualstudio.com' project 'f0f0a9fa-a2ad-47bc-b6f0-2c080f9a420f' was scanned.  (`not able to access required ADO API to check`). To support security of the pipeline operations, connections must not be granted access to all pipelines. This is in keeping with the principle of least privilege because a vulnerability in components used by one pipeline can be leveraged by an attacker to attack other pipelines having access to critical resources. 1. Navigate to the service connections settings page for your project --> 2. Select your service connection from the list --> 3. Click on three dots --> 4. Click Security --> 5. Under 'Pipeline Permissions', click 'Restrict Permission' to avoid granting access to all pipelines and add only those pipelines that need explicit access on service connection.

without the message:
[Service connection Id:AZC102.serviceconnection.DotNotShareWithAllPipelines.json]() in organization 'testorg.visualstudio.com' project 'f0f0a9fa-a2ad-47bc-b6f0-2c080f9a420f' was scanned.` `. To support security of the pipeline operations, connections must not be granted access to all pipelines. This is in keeping with the principle of least privilege because a vulnerability in components used by one pipeline can be leveraged by an attacker to attack other pipelines having access to critical resources. 1. Navigate to the service connections settings page for your project --> 2. Select your service connection from the list --> 3. Click on three dots --> 4. Click Security --> 5. Under 'Pipeline Permissions', click 'Restrict Permission' to avoid granting access to all pipelines and add only those pipelines that need explicit access on service connection.

pls let me know if what do you think?